File: swift/Tests/RemitMdTests/ComplianceTests.swift
import XCTest
import Foundation
#if canImport(FoundationNetworking)
import FoundationNetworking
#endif
@testable import RemitMd

/// Compliance tests: Swift SDK against a real running server.
///
/// Tests are skipped when the server is not reachable. Boot the server with:
///   docker compose -f docker-compose.compliance.yml up -d
///
/// Environment variables:
///   REMIT_TEST_SERVER_URL  (default: http://localhost:3000)
///   REMIT_ROUTER_ADDRESS   (default: 0x70997970C51812dc3A010C7d01b50e0d17dc79C8)
final class ComplianceTests: XCTestCase {

    // MARK: - Config

    private static let serverURL: String = {
        ProcessInfo.processInfo.environment["REMIT_TEST_SERVER_URL"] ?? "http://localhost:3000"
    }()

    private static let routerAddress: String = {
        ProcessInfo.processInfo.environment["REMIT_ROUTER_ADDRESS"]
            ?? "0x70997970C51812dc3A010C7d01b50e0d17dc79C8"
    }()

    private static var _serverAvailable: Bool?
    private static var sharedPayerKey: String?
    private static var sharedPayerAddr: String?

    // MARK: - Availability

    private static func isServerAvailable() async -> Bool {
        if let cached = _serverAvailable { return cached }
        guard let url = URL(string: serverURL + "/health") else {
            print("[COMPLIANCE] server availability check: invalid URL")
            _serverAvailable = false
            return false
        }
        var req = URLRequest(url: url, timeoutInterval: 3)
        req.httpMethod = "GET"
        do {
            let (_, resp) = try await URLSession.shared.data(for: req)
            let ok = (resp as? HTTPURLResponse)?.statusCode == 200
            _serverAvailable = ok
            print("[COMPLIANCE] server availability: \(ok ? "REACHABLE" : "UNREACHABLE") at \(serverURL)")
            return ok
        } catch {
            _serverAvailable = false
            print("[COMPLIANCE] server availability: UNREACHABLE at \(serverURL) error=\(error.localizedDescription)")
            return false
        }
    }

    // MARK: - Helpers

    /// Makes a JSON POST/GET and decodes the response body.
    private static func httpJSON(
        method: String,
        path: String,
        body: [String: Any]? = nil,
        bearerToken: String? = nil
    ) async throws -> [String: Any] {
        let url = URL(string: serverURL + path)!
        var req = URLRequest(url: url, timeoutInterval: 10)
        req.httpMethod = method
        req.setValue("application/json", forHTTPHeaderField: "Content-Type")
        if let token = bearerToken {
            req.setValue("Bearer \(token)", forHTTPHeaderField: "Authorization")
        }
        if let body = body {
            req.httpBody = try JSONSerialization.data(withJSONObject: body)
        }
        let (data, _) = try await URLSession.shared.data(for: req)
        return try JSONSerialization.jsonObject(with: data) as! [String: Any]
    }

    /// Generate a random private key and derive the wallet address.
    private static func generateWallet() throws -> (String, String) {
        let keyBytes = (0..<32).map { _ in UInt8.random(in: 0...255) }
        let privateKey = "0x" + keyBytes.map { String(format: "%02x", $0) }.joined()
        let wallet = try makeWallet(privateKey: privateKey)
        print("[COMPLIANCE] wallet created: \(wallet.address) (chain=84532)")
        return (privateKey, wallet.address)
    }

    /// Fund a wallet via mint (no auth required in testnet mode).
    private static func fundWallet(_ walletAddr: String) async throws {
        print("[COMPLIANCE] mint: requesting 1000 USDC -> \(walletAddr)")
        let resp = try await httpJSON(
            method: "POST",
            path: "/api/v1/mint",
            body: ["wallet": walletAddr, "amount": 1000]
        )
        guard let txHash = resp["tx_hash"] as? String else {
            print("[COMPLIANCE] mint: FAILED for \(walletAddr) response=\(resp)")
            throw NSError(domain: "Compliance", code: 3,
                          userInfo: [NSLocalizedDescriptionKey: "mint failed: \(resp)"])
        }
        print("[COMPLIANCE] mint: 1000 USDC -> \(walletAddr) tx=\(txHash)")
    }

    private static func makeWallet(privateKey: String) throws -> RemitWallet {
        try RemitWallet(
            privateKey: privateKey,
            chain: .baseSepolia,
            baseURL: serverURL,
            routerAddress: routerAddress
        )
    }

    private static func getSharedPayer() async throws -> RemitWallet {
        if let key = sharedPayerKey {
            print("[COMPLIANCE] reusing shared payer: \(sharedPayerAddr ?? "unknown")")
            return try makeWallet(privateKey: key)
        }
        let (pk, addr) = try generateWallet()
        try await fundWallet(addr)
        sharedPayerKey  = pk
        sharedPayerAddr = addr
        print("[COMPLIANCE] shared payer initialized: \(addr)")
        return try makeWallet(privateKey: pk)
    }

    // MARK: - Auth tests

    func testCompliance_AuthenticatedRequest_ReturnsBalance_Not401() async throws {
        print("[COMPLIANCE] === test: AuthenticatedRequest_ReturnsBalance_Not401 ===")
        guard await Self.isServerAvailable() else {
            print("[COMPLIANCE] SKIPPED: server not available")
            return
        }

        let (pk, _) = try Self.generateWallet()
        let wallet = try Self.makeWallet(privateKey: pk)
        print("[COMPLIANCE] requesting balance for \(wallet.address)")

        // balance() makes an authenticated GET - will throw on 401.
        let balance = try await wallet.balance()
        print("[COMPLIANCE] balance: \(balance.balance) \(balance.currency) for \(balance.address)")
        XCTAssertNotNil(balance)
    }

    func testCompliance_UnauthenticatedRequest_Returns401() async throws {
        print("[COMPLIANCE] === test: UnauthenticatedRequest_Returns401 ===")
        guard await Self.isServerAvailable() else {
            print("[COMPLIANCE] SKIPPED: server not available")
            return
        }

        let url = URL(string: Self.serverURL + "/api/v1/payments/direct")!
        var req = URLRequest(url: url, timeoutInterval: 5)
        req.httpMethod = "POST"
        req.setValue("application/json", forHTTPHeaderField: "Content-Type")
        req.httpBody = try JSONSerialization.data(withJSONObject: [
            "to": "0x70997970C51812dc3A010C7d01b50e0d17dc79C8",
            "amount": "1.000000",
        ])
        print("[COMPLIANCE] sending unauthenticated POST to /api/v1/payments/direct")
        let (_, resp) = try await URLSession.shared.data(for: req)
        let statusCode = (resp as? HTTPURLResponse)?.statusCode ?? 0
        print("[COMPLIANCE] unauthenticated response: status=\(statusCode)")
        XCTAssertEqual(statusCode, 401,
            "unauthenticated POST must return 401")
    }

    // MARK: - Payment tests

    func testCompliance_PayDirect_HappyPath_ReturnsTxHash() async throws {
        print("[COMPLIANCE] === test: PayDirect_HappyPath_ReturnsTxHash ===")
        guard await Self.isServerAvailable() else {
            print("[COMPLIANCE] SKIPPED: server not available")
            return
        }

        let payer = try await Self.getSharedPayer()
        let (_, payeeAddr) = try Self.generateWallet()
        print("[COMPLIANCE] pay: 5.0 USDC \(payer.address) -> \(payeeAddr) memo=\"swift compliance test\"")

        let tx = try await payer.pay(to: payeeAddr, amount: 5.0, memo: "swift compliance test")
        print("[COMPLIANCE] pay: 5.0 USDC \(payer.address) -> \(payeeAddr) tx=\(tx.txHash ?? "nil") invoice=\(tx.invoiceId ?? "nil") status=\(tx.status)")

        XCTAssertFalse(tx.txHash?.isEmpty ?? true,
            "pay() must return a non-empty txHash")
    }

    func testCompliance_PayDirect_BelowMinimum_ThrowsError() async throws {
        print("[COMPLIANCE] === test: PayDirect_BelowMinimum_ThrowsError ===")
        guard await Self.isServerAvailable() else {
            print("[COMPLIANCE] SKIPPED: server not available")
            return
        }

        let payer = try await Self.getSharedPayer()
        let (_, payeeAddr) = try Self.generateWallet()
        print("[COMPLIANCE] pay: 0.0001 USDC \(payer.address) -> \(payeeAddr) (expecting error)")

        do {
            _ = try await payer.pay(to: payeeAddr, amount: 0.0001)
            XCTFail("pay() with amount below minimum must throw an error")
        } catch {
            print("[COMPLIANCE] pay below minimum: correctly threw error: \(error)")
            // Expected - any error is acceptable.
        }
    }
}

File: dotnet/RemitMd.Tests/ComplianceTests.cs
using System;
using System.Net.Http;
using System.Net.Http.Json;
using System.Text.Json;
using System.Threading;
using System.Threading.Tasks;
using Xunit;
using Xunit.Abstractions;

namespace RemitMd.Tests;

/// <summary>
/// Compliance tests: .NET SDK against a real running server.
///
/// Tests return early (pass vacuously) when the server is not reachable.
/// Boot the server with:
///   docker compose -f docker-compose.compliance.yml up -d
///
/// Environment variables:
///   REMIT_TEST_SERVER_URL  (default: http://localhost:3000)
///   REMIT_ROUTER_ADDRESS   (default: 0x70997970C51812dc3A010C7d01b50e0d17dc79C8)
/// </summary>
public class ComplianceTests
{
    private static readonly string ServerUrl =
        Environment.GetEnvironmentVariable("REMIT_TEST_SERVER_URL") ?? "http://localhost:3000";
    private static readonly string RouterAddress =
        Environment.GetEnvironmentVariable("REMIT_ROUTER_ADDRESS") ??
        "0x70997970C51812dc3A010C7d01b50e0d17dc79C8";

    private static readonly HttpClient Http = new()
    {
        Timeout = TimeSpan.FromSeconds(10),
    };

    private static readonly SemaphoreSlim PayerLock = new(1, 1);
    private static Wallet? _sharedPayer;

    // ─── Server availability ──────────────────────────────────────────────────

    private static readonly Lazy<Task<bool>> ServerAvailable = new(async () =>
    {
        try
        {
            using var cts = new CancellationTokenSource(TimeSpan.FromSeconds(3));
            var resp = await Http.GetAsync(ServerUrl + "/health", cts.Token);
            return resp.IsSuccessStatusCode;
        }
        catch
        {
            return false;
        }
    });

    // ─── Helpers ──────────────────────────────────────────────────────────────

    private static (string PrivateKey, string WalletAddress) GenerateWallet()
    {
        var keyBytes = System.Security.Cryptography.RandomNumberGenerator.GetBytes(32);
        var privateKey = "0x" + Convert.ToHexString(keyBytes).ToLowerInvariant();
        var wallet = MakeWallet(privateKey);
        Console.WriteLine($"[COMPLIANCE] wallet created: {wallet.Address} (chain=84532)");
        return (privateKey, wallet.Address);
    }

    private static async Task FundWallet(string walletAddress)
    {
        Console.WriteLine($"[COMPLIANCE] minting 1000 USDC to {walletAddress} ...");
        var resp = await Http.PostAsJsonAsync($"{ServerUrl}/api/v1/mint", new
        {
            wallet = walletAddress,
            amount = 1000,
        });
        resp.EnsureSuccessStatusCode();
        var body = await resp.Content.ReadAsStringAsync();
        using var doc = JsonDocument.Parse(body);
        Assert.True(doc.RootElement.TryGetProperty("tx_hash", out var txHashElem),
            $"mint response must contain tx_hash");
        Console.WriteLine($"[COMPLIANCE] mint: 1000 USDC -> {walletAddress} tx={txHashElem}");
    }

    private static Wallet MakeWallet(string privateKey) =>
        new Wallet(privateKey, chain: "base", testnet: true,
            baseUrl: ServerUrl, routerAddress: RouterAddress);

    private async Task<Wallet> GetSharedPayer()
    {
        if (_sharedPayer is not null) return _sharedPayer;

        await PayerLock.WaitAsync();
        try
        {
            if (_sharedPayer is not null) return _sharedPayer;

            Console.WriteLine("[COMPLIANCE] creating shared payer wallet ...");
            var (pk, addr) = GenerateWallet();
            await FundWallet(addr);
            _sharedPayer = MakeWallet(pk);
            Console.WriteLine($"[COMPLIANCE] shared payer ready: {_sharedPayer.Address}");
        }
        finally
        {
            PayerLock.Release();
        }
        return _sharedPayer!;
    }

    // ─── Auth tests ───────────────────────────────────────────────────────────

    [Fact]
    public async Task Compliance_AuthenticatedRequest_ReturnsBalance_Not401()
    {
        if (!await ServerAvailable.Value)
        {
            Console.WriteLine("[COMPLIANCE] server not reachable — skipping auth test");
            return;
        }

        Console.WriteLine("[COMPLIANCE] === AuthenticatedRequest_ReturnsBalance_Not401 ===");
        var (pk, _) = GenerateWallet();
        var wallet = MakeWallet(pk);

        // BalanceAsync() makes an authenticated GET - will throw on 401.
        Console.WriteLine($"[COMPLIANCE] fetching balance for {wallet.Address} ...");
        var balance = await wallet.BalanceAsync();
        Assert.NotNull(balance);
        Console.WriteLine($"[COMPLIANCE] balance: {balance.Usdc} USDC (address={balance.Address}, chainId={balance.ChainId})");
    }

    [Fact]
    public async Task Compliance_UnauthenticatedRequest_Returns401()
    {
        if (!await ServerAvailable.Value)
        {
            Console.WriteLine("[COMPLIANCE] server not reachable — skipping 401 test");
            return;
        }

        Console.WriteLine("[COMPLIANCE] === UnauthenticatedRequest_Returns401 ===");
        Console.WriteLine("[COMPLIANCE] sending unauthenticated POST to /api/v1/payments/direct ...");
        var resp = await Http.PostAsJsonAsync($"{ServerUrl}/api/v1/payments/direct", new
        {
            to     = "0x70997970C51812dc3A010C7d01b50e0d17dc79C8",
            amount = "1.000000",
        });
        Console.WriteLine($"[COMPLIANCE] response status: {(int)resp.StatusCode}");
        Assert.Equal(401, (int)resp.StatusCode);
        Console.WriteLine("[COMPLIANCE] confirmed: unauthenticated request returned 401");
    }

    // ─── Payment tests ────────────────────────────────────────────────────────

    [Fact]
    public async Task Compliance_PayDirect_HappyPath_ReturnsTxHash()
    {
        if (!await ServerAvailable.Value)
        {
            Console.WriteLine("[COMPLIANCE] server not reachable — skipping pay direct test");
            return;
        }

        Console.WriteLine("[COMPLIANCE] === PayDirect_HappyPath_ReturnsTxHash ===");
        var payer = await GetSharedPayer();
        var (_, payeeAddr) = GenerateWallet();

        Console.WriteLine($"[COMPLIANCE] pay: 5.0 USDC {payer.Address} -> {payeeAddr} ...");
        var tx = await payer.PayAsync(payeeAddr, 5.0m, memo: "dotnet compliance test");

        Assert.False(string.IsNullOrEmpty(tx.TxHash),
            "PayAsync() must return a non-empty TxHash");
        Console.WriteLine($"[COMPLIANCE] pay: 5.0 USDC {payer.Address} -> {payeeAddr} tx={tx.TxHash} id={tx.Id}");
    }

    [Fact]
    public async Task Compliance_PayDirect_BelowMinimum_ThrowsRemitError()
    {
        if (!await ServerAvailable.Value)
        {
            Console.WriteLine("[COMPLIANCE] server not reachable — skipping below-minimum test");
            return;
        }

        Console.WriteLine("[COMPLIANCE] === PayDirect_BelowMinimum_ThrowsRemitError ===");
        var payer = await GetSharedPayer();
        var (_, payeeAddr) = GenerateWallet();

        Console.WriteLine($"[COMPLIANCE] pay: 0.0001 USDC {payer.Address} -> {payeeAddr} (expect error) ...");
        var ex = await Assert.ThrowsAsync<RemitError>(async () =>
            await payer.PayAsync(payeeAddr, 0.0001m));
        Console.WriteLine($"[COMPLIANCE] correctly threw RemitError: {ex.Message}");
    }
}
